README.md
### Underground Nexus deployment
> Current version 1.0.0

To pull the latest Underground Nexus image.

`docker pull pyrrhus/nexus0:latest`

**Resources**
1. PI-Hole DNS that sits behind NGINX Reverse Proxy - Obtain the IP address of your running docker container. `docker inspect Underground-Nexus`
  - There is no password set, so you will have to change it
    - Access the Docker container once you're within the Underground-Nexus running container.
      - `docker exec -it Inner-DNS-Control /bin/sh`
      - `sudo pihole -a -p`
2. Linux webtop desktop - Alpine KDE - `http://10.20.0.30:3000` (optionally, only accessible from the Ubuntu MATE desktop)
3. Linux webtop desktop - Ubuntu MATE - `http://172.17.0.4:1000`
4. Kali Linux Bleeding Edge Repository - not publicly accessible
5. Portainer CE - `https://172.17.0.4:9443/`
6. Minio S3 Storage - `http://172.17.0.4:9001/login`
  - To access Minio S3, this server is deployed with the default credentials: `minioadmin:minioadmin`
7. Visual Studio Code - `http://10.20.0.73:3000` is not publicly accessible, this can be accessed from Ubuntu MATE desktop.
8. Kubernetes (k3d) cluster deployed on standby as a load-balancer

**Features that are missing:**

UPDATE: 
  - Added Visual Studio Code, there were issues with hardware acceleration when it is deployed via ARM64 distro.
  - Minio has been added back again, and the port has been fixed.
  - ~~There is a bash script now, and I took away the workbench.sh script for now.~~ Scratch that, it's added back in the Dockerfile.
  - Docker Swarm is initialized last at the end of the script to avoid conflicts with existing services.

1. ~~Minio is not being used right now because Minio has a conflicting port 9000 with Portainer. Future plan to redirect port usage or take away port 9000 on Portainer.~~
2. There isn't the best bash script to check for logic on when there are certain resources that are already available. A cli tool is coming soon.
3. Docker Swarm is not initialized first only because of the conflict that Portainer has while being deployed along with Docker Swarm turned on.

**ROADMAP**

1. Before, I generated `cosign` key and `cosign` public key, now it has been changed to use GitHub support. There is no local private key to verify. At this time, signing and verification is done with GitHub OIDC login.
2. Build an arm64 version of Underground Nexus.
3. Dagger plan is using `docker#Push`, looking for a better way to implement this steps through Github Actions.
4. So far, docker images are being built and pushed to DockerHub registry. Future plan is create an in-house self-hosted registry.

**Information**

The default Dockerfile has the original `docker:dind` base image to build the Underground Nexus from. The GitHub Action will build the main Dockerfile with `dagger`. Please note, there are two Dockerfiles. One is using `nestybox/alpine-supervisord-docker:latest` which is an alternate version of Docker-in-Docker (dind).

Before, I had multiple steps to build and load the newly built Docker images. I was able to change the Dagger actions to build with the Dockerfile contents.

To start, you will need to set some variables.

```bash
export REGISTRY_DOCKERIO_USER=<USERNAME>
export REGISTRY_DOCKERIO_PASS=<PASSWORD>
```

To build with the latest version or a specific version. Once built, it will push to Dockerhub.

```bash
dagger do versions <VERSION>
```

To run with just Docker in `--privileged` mode.

```bash
docker run -itd --name=Underground-Nexus \
    -h Underground-Nexus \
    --privileged \
    --init \
    -p 1000:1000 -p 9050:9443 \
    -v underground-nexus-docker-socket:/var/run \
    -v underground-nexus-data:/var/lib/docker/volumes \
    -v nexus-bucket:/nexus-bucket pyrrhus/nexus0:latest
```

Execute the `deploy-olympiad.sh` script

`docker exec Underground-Nexus sh deploy-olympiad.sh`

If you need to troubleshoot or access the running container. Do this below.

`docker exec -it Underground-Nexus /bin/sh`

**Details**

I am using Sigstore Cosign to generate key-pair and sign my docker images with the generated Cosign private key.

Cosign public key

```bash
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEKEBgXjBv494co5Ko5KED3BDrC1+Z
v8gRvhT0mhF8anj95LSFVoSSfptc/oOQGVk4m+B24aCmYURIDCyaI8i6Kw==
-----END PUBLIC KEY-----
```

To verify the signed image with the about Cosign public key. You can also download `crane` to display the signed digest. It is also verifiable on my Dockerhub page.

```bash
docker pull pyrrhus/nexus0:latest

cosign triangulate pyrrhus/nexus0:latest
index.docker.io/pyrrhus/nexus0:sha256-2370cc6b72bbeb73eab1f2edf3e22a5e30c785877b82a5d2ea79302742243927.sig

crane manifest $(cosign triangulate pyrrhus/nexus0:latest) | jq .

{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "size": 233,
    "digest": "sha256:342957a565a3133b864b091291dca2778d9f2735c2fc31404937cf1076b029a7"
  },
  "layers": [
    {
      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
      "size": 319,
      "digest": "sha256:863fd2246616b5842ec6dfaefa40ba91acb5627d4d1a0e16e136f75f5e25eaa6",
      "annotations": {
        "dev.cosignproject.cosign/signature": "MEQCIDnIHpAtsWKlc4Tlu0aCK8d/8jyF0nUZjPfaGbSwwZJqAiAteCPKYuG4na8KktWj48CxEMjKdSleIQS0NzKDAsO5CA=="
      }
    }
  ]
}

cosign verify --key cosign.pub pyrrhus/nexus0:latest

Verification for index.docker.io/pyrrhus/nexus0:latest --
The following checks were performed on each of these signatures:
  - The signatures were verified against the specified public key

[{"critical":{"identity":{"docker-reference":"index.docker.io/pyrrhus/nexus0"},"image":{"docker-manifest-digest":"sha256:2370cc6b72bbeb73eab1f2edf3e22a5e30c785877b82a5d2ea79302742243927"},"type":"cosign container image signature"},"optional":{"commit":"fbc8d91acd7fac6a70a46d71d88b15b8550fe2d8","repo":"pyrrhus/nexus0"}}]
```

***Experimental***

Since I built this particular Docker image with `sysbox`, you can use `sysbox` as the runtime. To run with `sysbox`, make sure that `sysbox` is installed.

```bash
docker run --runtime=sysbox-runc \
    -itd \
    --name=Underground-Nexus \
    --hostname=Underground-Nexus \
    --mount source=underground-nexus-data,target=/var/lib/docker/volumes \
    --mount source=underground-nexus-docker-socket,target=/var/run \
    --mount source=nexus-bucket,target=/nexus-bucket \
    -p 1000:1000 -p 9050:9443 nexus0:v0.7.1
```

Using `syft` to package the newly built docker image.

```bash
syft packages pyrrhus/nexus0:latest -o spdx > nexus0-dind-latest.spdx

# Cosign attach sbom to first docker built image
cosign attach sbom --sbom nexus0-dind-latest.spdx pyrrhus/nexus0:latest
WARNING: Attaching SBOMs this way does not sign them. If you want to sign them, use 'cosign attest -predicate nexus0-dind-latest.spdx -key <key path>' or 'cosign sign -key <key path> <sbom image>'.
Uploading SBOM file for [index.docker.io/pyrrhus/nexus0:latest] to [index.docker.io/pyrrhus/nexus0:sha256-2370cc6b72bbeb73eab1f2edf3e22a5e30c785877b82a5d2ea79302742243927.sbom] with mediaType [text/spdx].

# Cosign sign sbom
cosign sign --key cosign.key index.docker.io/pyrrhus/nexus0:sha256-2370cc6b72bbeb73eab1f2edf3e22a5e30c785877b82a5d2ea79302742243927.sbom
Enter password for private key: 
Pushing signature to: index.docker.io/pyrrhus/nexus0

# Cosign verify
cosign verify --key cosign.pub index.docker.io/pyrrhus/nexus0:sha256-2370cc6b72bbeb73eab1f2edf3e22a5e30c785877b82a5d2ea79302742243927.sbom

Verification for index.docker.io/pyrrhus/nexus0:sha256-2370cc6b72bbeb73eab1f2edf3e22a5e30c785877b82a5d2ea79302742243927.sbom --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - The signatures were verified against the specified public key

[{"critical":{"identity":{"docker-reference":"index.docker.io/pyrrhus/nexus0"},"image":{"docker-manifest-digest":"sha256:fb55e6584d0c079eac55b4f58bb50809bc4438e31611715f0fac6579f764e0cf"},"type":"cosign container image signature"},"optional":null}]
```

You can also use `grype` to scan the new docker image.

```bash
grype pyrrhus/nexus0:latest --add-cpes-if-none     
 ✔ Vulnerability DB        [updated]
 ✔ Loaded image            
 ✔ Parsed image            
 ✔ Cataloged packages      [377 packages]
 ✔ Scanned image           [14 vulnerabilities]

NAME                              INSTALLED  FIXED-IN  TYPE       VULNERABILITY        SEVERITY 
docker                            5.0.3                python     CVE-2019-13139       High      
docker                            5.0.3                python     CVE-2020-27534       Medium    
docker                            5.0.3                python     CVE-2018-10892       Medium    
docker                            5.0.3                python     CVE-2019-13509       High      
docker                            5.0.3                python     CVE-2019-5736        High      
docker                            5.0.3                python     CVE-2021-21284       Medium    
docker                            5.0.3                python     CVE-2021-21285       Medium    
docker                            5.0.3                python     CVE-2019-16884       High      
github.com/containerd/containerd  v1.6.1     1.6.6     go-module  GHSA-5ffw-gxpp-mxpf  Medium    
github.com/opencontainers/runc    v1.1.0     1.1.2     go-module  GHSA-f3fp-gc8g-vw66  Medium    
google.golang.org/protobuf        v1.27.1              go-module  CVE-2015-5237        High      
google.golang.org/protobuf        v1.27.1              go-module  CVE-2021-22570       High 
```

CHANGELOG.md
v0.8.0

**Assets**:

- pyrrhus/nexus0:latest
- pyrrhus/nexus0:sha256-138dd9d7d21a53efc6b86a287e0d7fe86bd143fd3c673add5de796ba640579aa.sig

- pyrrhus/nexus0:v0.8.0-2583187478
- pyrrhus/nexus0:sha256-e41c0e44449eaf75a3a72d393762a35806c38de47755f61a7a298db4139c65aa.sig

- pyrrhus/nexus0:v0.8.0-2583187479-sysbox
- pyrrhus/nexus0:sha256-275681cfcf97f310bb5cbef9d1453ad6ca5d4a8644ee87d43851009ad9ecb000.sig

deploy-olympiad.sh
#!/bin/sh

set -e

export PIHOLE_IMAGE=pihole/pihole:latest
export PORTAINER_IMAGE=portainer/portainer-ce:2.14.0
export MINIO_IMAGE=quay.io/minio/minio
export VAULT_IMAGE=vault
export KALI_LINUX_IMAGE=registry.gitlab.com/cyberphxv/nexus-athena0:latest.amd64
export WORKBENCH_WEBTOP_IMAGE=pyrrhus/webtop-workbench:amd64-latest
export SOC_WEBTOP_IMAGE=pyrrhus/soc-admin-webtop:amd64-latest
export VISUAL_STUDIO_IMAGE=lscr.io/linuxserver/openvscode-server

echo "Creating volumes for persistence..."
docker volume create pihole_dns_data && \
    docker volume create portainer_data

echo "Pull necessary images"
docker pull $PIHOLE_IMAGE
docker pull $PORTAINER_IMAGE
docker pull $MINIO_IMAGE
docker pull $VAULT_IMAGE
docker pull $KALI_LINUX_IMAGE
docker pull $WORKBENCH_WEBTOP_IMAGE
docker pull $SOC_WEBTOP_IMAGE

echo "Create Docker network - Inner-Athena"

export DOCKER_NETWORK=Inner-Athena
export SUBNET_NETWORK=10.20.0.20/24

if docker network ls | grep $DOCKER_NETWORK;
then
    echo "Docker's network Inner-Athena is already created."
else
    docker network create -d bridge --subnet=$SUBNET_NETWORK $DOCKER_NETWORK 2> /dev/null
fi

sleep 5

echo "Deploy Pi-Hole DNS Server"

export PIHOLE=Inner-DNS-Control
export PIHOLE_IPADDRESS=10.20.0.20

if docker ps --format "{{.Names}}" | grep -w $PIHOLE;
then
    echo "Inner-DNS-Control is already created."
else
    docker compose -f docker-compose.yml up -d
fi

echo "Build Olympiad0 Portainer node"

sleep 30

export PORTAINER_NAME=Olympiad0

if docker volume ls | grep -w portainer_data;
then
    echo "Portainer data volume exists"
else
    docker compose -f portainer-deploy.yml up -d
fi

echo "Build Cyber Life Torpedo"

sleep 5

export MINIO_NAME=torpedo

if docker ps --format "{{.Names}}" | grep -w $MINIO_NAME;
then
    echo "torpedo is already created."
else
    docker run \
        -itd \
        --privileged \
        -p 9000:9000 \
        -p 9001:9001 \
        --name=$MINIO_NAME \
        --hostname $MINIO_NAME \
        --dns=$PIHOLE_IPADDRESS \
        --net=$DOCKER_NETWORK \
        --restart=always \
        -v /nexus-bucket:/nexus-bucket \
        -v /nexus-bucket/s3-torpedo:/data $MINIO_IMAGE server /data --console-address ":9001" 2> /dev/null
fi

echo "Build Development Vault server"

sleep 5

export SECRET_VAULT_NAME=Nexus-Secret-Vault

if docker ps --format "{{.Names}}" | grep -w $SECRET_VAULT_NAME;
then
    echo "Nexus Secret Vault is already created."
else
    docker run \
        -itd -p 8200:1234 \
        --name=$SECRET_VAULT_NAME \
        --hostname $SECRET_VAULT_NAME \
        --dns=$PIHOLE_IPADDRESS \
        --net=$DOCKER_NETWORK \
        --restart=always \
        --cap-add=IPC_LOCK \
        -e 'VAULT_DEV_ROOT_TOKEN_ID=myroot' -e 'VAULT_DEV_LISTEN_ADDRESS=0.0.0.0:1234' $VAULT_IMAGE 2> /dev/null
fi

echo "Build Athena0 Stack"

sleep 5

export KALI_LINUX_NAME=Athena0

if docker ps --format "{{.Names}}" | grep -w $KALI_LINUX_NAME;
then
    echo "Athena0 is already created."
else
    docker run \
        -itd \
        -p 22:22 \
        --name=$KALI_LINUX_NAME \
        --hostname $KALI_LINUX_NAME \
        --dns=$PIHOLE_IPADDRESS \
        --net=$DOCKER_NETWORK \
        --restart=always \
        -v athena0:/home/ \
        -v /nexus-bucket:/nexus-bucket \
        -v /etc/docker:/etc/docker \
        -v /usr/local/bin/docker:/usr/local/bin/docker \
        -v /var/run/docker.sock:/var/run/docker.sock $KALI_LINUX_IMAGE 2> /dev/null
fi

echo "Build Workbench"

sleep 5

export WORKBENCH_NAME=workbench

if docker ps --format "{{.Names}}" | grep -w $WORKBENCH_NAME;
then
    echo "workbench is already created."
else
    docker run \
        -itd \
        --name=$WORKBENCH_NAME \
        --hostname $WORKBENCH_NAME \
        --privileged \
        --init \
        -e PUID=1000 \
        -e PGID=1000 \
        -e TZ=America/Chicago \
        -p 1000:3000 \
        --dns=$PIHOLE_IPADDRESS \
        --net=$DOCKER_NETWORK \
        --restart=always \
        -v workbench0:/config \
        -v /nexus-bucket:/config/Desktop/nexus-bucket \
        -v /var/run/docker.sock:/var/run/docker.sock $WORKBENCH_WEBTOP_IMAGE 2> /dev/null
fi

echo "Build Security Operation Center"

sleep 5

export SOC_NAME=Security-Operation-Center
export SOC_IPADDRESS=10.20.0.30

if docker ps --format "{{.Names}}" | grep -w $SOC_NAME;
then
    echo "Security Operation Center is already created."
else
    docker run \
        -itd \
        --name=$SOC_NAME \
        --hostname $SOC_NAME \
        --privileged \
        --init \
        -e PUID=2000 \
        -e PGID=2000 \
        -e TZ=America/Chicago \
        -p 2000:3000 \
        --dns=$PIHOLE_IPADDRESS \
        --net=$DOCKER_NETWORK \
        --ip=$SOC_IPADDRESS \
        --restart=always \
        -v security-operation-center:/config \
        -v /nexus-bucket:/config/Desktop/nexus-bucket $SOC_WEBTOP_IMAGE 2> /dev/null
fi

echo "Install Visual Studio"

export VISUALSTUDIO_NAME=code-server

if docker ps --format "{{.Names}}" | grep -w $VISUALSTUDIO_NAME;
then
    echo "Security Operation Center is already created."
else
    docker run \
        -d \
        --name=$VISUALSTUDIO_NAME \
        -e PUID=1050 \
        -e PGID=1050 \
        -p 18443:3000 \
        --dns=$PIHOLE_IPADDRESS \
        --net=$DOCKER_NETWORK \
        -v /nexus-bucket:/nexus-bucket \
        -v /nexus-bucket/visual-studio-code:/config \
        -v /etc/docker:/etc/docker \
        -v /usr/local/bin/docker:/usr/local/bin/docker \
        -v /var/run/docker.sock:/var/run/docker.sock \
        --restart unless-stopped $VISUAL_STUDIO_IMAGE 2> /dev/null
fi

echo "Intiating Docker Swarm"

sleep 3; docker swarm init

wget https://raw.githubusercontent.com/Underground-Ops/underground-nexus/main/Production%20Artifacts/firefox--hostnameomepage.sh && \
    sh firefox--hostnameomepage.sh

sleep 5

sh ./nexus-bucket/workbench.sh

echo "Deploy KuberNexus"

# Install Kubernetes on AMD64
echo "Install Kubectl"

curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" && install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl

sleep 5

wget -q -O - https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash

k3d cluster create KuberNexus \
    -p 8080:80@loadbalancer \
    -p 8443:8443@loadbalancer \
    -p 2222:22@loadbalancer \
    -p 179:179@loadbalancer \
    -p 2375:2376@loadbalancer \
    -p 2378:2379@loadbalancer \
    -p 2381:2380@loadbalancer \
    -p 8472:8472@loadbalancer \
    -p 8843:443@loadbalancer \
    -p 4789:4789@loadbalancer \
    -p 9099:9099@loadbalancer \
    -p 9100:9100@loadbalancer \
    -p 7443:9443@loadbalancer \
    -p 9796:9796@loadbalancer \
    -p 6783:6783@loadbalancer \
    -p 10250:10250@loadbalancer \
    -p 10254:10254@loadbalancer \
    -p 31896:31896@loadbalancer

docker compose -f portainer-deploy.yml up -d

echo "Completed."